setup.py

```diff
@@ -3,128 +3,78 @@
-necessary to support the oldest and buggiest of compiler specimens. Now that
-C++11-compatible compilers are widely available, this heavy machinery has
-become an excessively large and unnecessary dependency.
-
-Think of this library as a tiny self-contained version of Boost.Python with
-everything stripped away that isn't relevant for binding generation. Without
-comments, the core header files only require ~4K lines of code and depend on
-Python (2.7 or 3.x, or PyPy2.7 >= 5.7) and the C++ standard library. This
-compact implementation was possible thanks to some of the new C++11 language
-features (specifically: tuples, lambda functions and variadic templates). Since
-its creation, this library has grown beyond Boost.Python in many ways, leading
-to dramatically simpler binding code in many common situations.""")
+# PYBIND11_ALT_SDIST will build a different sdist, with the python-headers
+# files, and the sys.prefix files (CMake and headers).
+
+alt_sdist = os.environ.get("PYBIND11_ALT_SDIST", False)
+setup_py = "tools/setup_alt.py" if alt_sdist else "tools/setup_main.py"
+pyproject_toml = "tools/pyproject.toml"
+
+# In a PEP 518 build, this will be in its own environment, so it will not
+# create extra files in the source
+
+DIR = os.path.abspath(os.path.dirname(__file__))
+
+
+@contextlib.contextmanager
+def monkey_patch_file(input_file, replacement_file):
+    "Allow a file to be temporarily replaced"
+    inp_file = os.path.abspath(os.path.join(DIR, input_file))
+    rep_file = os.path.abspath(os.path.join(DIR, replacement_file))
+
+    with open(inp_file, "rb") as f:
+        contents = f.read()
+    with open(rep_file, "rb") as f:
+        replacement = f.read()
+    try:
+        with open(inp_file, "wb") as f:
+            f.write(replacement)
+        yield
+    finally:
+        with open(inp_file, "wb") as f:
+            f.write(contents)
+
+
+@contextlib.contextmanager
+def TemporaryDirectory():  # noqa: N802
+    "Prepare a temporary directory, cleanup when done"
+    try:
+        tmpdir = tempfile.mkdtemp()
+        yield tmpdir
+    finally:
+        shutil.rmtree(tmpdir)
+
+
+@contextlib.contextmanager
+def remove_output(*sources):
+    try:
+        yield
+    finally:
+        for src in sources:
+            shutil.rmtree(src)
+
+
+with remove_output("pybind11/include", "pybind11/share"):
+    # Generate the files if they are not present.
+    with TemporaryDirectory() as tmpdir:
+        cmd = ["cmake", "-S", ".", "-B", tmpdir] + [
+            "-DCMAKE_INSTALL_PREFIX=pybind11",
+            "-DBUILD_TESTING=OFF",
+            "-DPYBIND11_NOPYTHON=ON",
+        ]
+        cmake_opts = dict(cwd=DIR, stdout=sys.stdout, stderr=sys.stderr)
+        subprocess.check_call(cmd, **cmake_opts)
+        subprocess.check_call(["cmake", "--install", tmpdir], **cmake_opts)
+
+    with monkey_patch_file("pyproject.toml", pyproject_toml):
+        with monkey_patch_file("setup.py", setup_py):
+            with open(setup_py) as f:
```

Comment:
These can all be combined, I think, rather than nesting:

```py
    with monkey_patch_file("pyproject.toml", pyproject_toml), \
         monkey_patch_file("setup.py", setup_py), \
         open(setup_py) as f:
```

Reply:
Not on Python 2.

Reply:
In fact, most of the useful context manager tools (`ExitStack`, `nullcontext`, `suppress`, `ExitStack`, `redirect_stdout`/`stderr`, and `ExitStack`) are also Python 3 only (or a backport). Did I mention `ExitStack`? :'(

Reply:
Huh, I tried this on Python 2 and it worked, I though?

Reply:
```
>>> import sys
>>> sys.version_info
sys.version_info(major=2, minor=7, micro=17, releaselevel='final', serial=0)
>>> with open("x", 'w'), \
...      open("y", 'w') as f:
...     print(f)
... 
<open file 'y', mode 'w' at 0x7fc13ed47d20>
```

`contextlib.nested` also seems to exist on 2, btw :-)

Reply:
(I also had my doubts and not too much hope, before trying)

Reply:
I’m wrong! It was New in 3 but backported to 2.7.  :D

Reply:
Another huge exercise in bikeshedding, though, but ... it does look better? ;-)

Reply:
I tried it, but I actually like the original nested one better, though - I really don't like ending lines with `\` (and Black doesn't either, though this would be a case where `# fmt: off` would be possible if we did want it), and the content is very short, so it isn't running into line length issues. And, due to the old parser, the ideal syntax:

```python
with (
        monkey_patch_file("pyproject.toml", pyproject_toml),
        monkey_patch_file("setup.py", setup_py),
        open(setup_py) as f
    ):
```

will not be available till Python 3.10.

Reply:
What do you think about this:

```python
    monkey_pyproject = monkey_patch_file("pyproject.toml", pyproject_toml)
    monkey_setup = monkey_patch_file("setup.py", setup_py)

    with monkey_pyproject, monkey_setup, open(setup_py) as f:
        code = compile(f.read(), setup_py, "exec")
        exec(code, globals(), {})
```

Reply:
PS: in Python 3, this is what I'd do:

```python
with ExitStack() as stack:
    stack.enter_context(monkey_patch_file("pyproject.toml", pyproject_toml))
    stack.enter_context(monkey_patch_file("setup.py", setup_py))
    f = stack.enter_context(open(setup_py))
```

Reply:
Right, OK. Whichever you prefer, then. I personally don't mind the `\` in this case, but it is quite tricky, yes.
Definitely not a blocking remarks!

Reply:
All gone now, using the proper hook. :)